ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -224,6 +229,25 @@ func (r *RayServiceReconciler) Reconcile(ctx context.Context, request ctrl.Reque
 	return ctrl.Result{RequeueAfter: ServiceDefaultRequeueDuration}, nil
 }
 
+func (r *RayServiceReconciler) calculateStatus(ctx context.Context, rayServiceInstance *rayv1.RayService, rayClusterInstance *rayv1.RayCluster) error {
+	serveSvc, err := common.BuildServeServiceForRayService(ctx, *rayServiceInstance, *rayClusterInstance)
+	if err != nil {
+		return err
+	}
+	serveEndPoints := &corev1.Endpoints{}
+	if err := r.Get(ctx, client.ObjectKey{Name: serveSvc.Name, Namespace: serveSvc.Namespace}, serveEndPoints); err != nil && !errors.IsNotFound(err) {
+		r.Log.Error(err, "Fail to retrieve the Kubernetes Endpoints from the cluster!")
+		return err
+	}
+
+	numServeEndpoints := 0
+	for _, subset := range serveEndPoints.Subsets {
```

Comment:
I am not familiar with endpoints. What is `Subsets`?

Reply:
These links explain it:  https://pkg.go.dev/k8s.io/api/core/v1#Endpoints
https://pkg.go.dev/k8s.io/api/core/v1#EndpointSubset